### ISSUE TITLE
Group by on content query web part not working as expected!

### ISSUE THREAD

**Issue by MacBoiler** [SharePoint/sp-dev-fx-webparts#1029]
Guys...can't be bothered filling out your big form...I need a group by handlebar template on the content query web part.

Make it so.

PS. I've tried adding one but I'm really out of my depth here and I've kind of promised my employer that I can do all this shit so please...I'm begging you...add a group by option to the content query web part.

**Comment by spplante** [comment]
@MacBoiler as a workaround, you can group by items by creating your own groupBy handlebar helper. Here is an example of a simple groupBy implementation I did for a quick demo, which supports grouping on text/date/url fields :

Refer the following MyExentensions.js file in the custom scripts section of the CQWP with the following code :

```javascript
ReactContentQuery.ExternalScripts.MyExtensions = {
    
    /****************************************************************************************************
     * Called BEFORE Handlebars has compiled and rendered the template 
     * @param wpContext : The context of the WebPart responsible for calling this callback
     * @param handlebarsContext : The Handlebars' instance associated with the WebPart  
     ****************************************************************************************************/
    onPreRender: function(wpContext, handlebarsContext) {
        
        // Registers the "groupBy" helper
        handlebarsContext.registerHelper('groupBy', function(items, field, options) {
            var groupedItems = [];
            
            items.forEach(function(item) {
                var groupKey = ReactContentQuery.ExternalScripts.MyExtensions.getGroupKey(item, field);
                var existingGroups = groupedItems.filter((group) => { return group.key == groupKey });
        
                if(existingGroups.length > 0) {
                    existingGroups[0].items.push(item);
                }
                else {
                    groupedItems.push({ key: groupKey, items: [item] });
                }
            }); 

            return groupedItems;
        });
    },


    /****************************************************************************************************
     * Returns the group key based on the given item and its field value
     * @param item : The item from which the group key must be extracted
     * @param field : The field/property of the item which must be used to extract the key  
     ****************************************************************************************************/
    getGroupKey: function(item, field) {
        
        // If item has no value for specified field return null
        if(!item[field]) {
            return null;
        }
    
        // If it's a date, returns the date without the time as a key
        else if(!isNaN(new Date(item[field].rawValue))) {
            var objDate = new Date(item[field].rawValue);
            objDate.setHours(0,0,0,0);
            return objDate.toISOString();
        }
    
        // If it's an url, returns the url as a key
        else if(item[field].rawValue.Url) {
            return item[field].rawValue.Url;
        }
    
        // If it's anything else, return the textValue
        else {
            return item[field].textValue;
        }
    }
    
}
```

Then in the handlebars template use as following :

```handlebars
<!-- Template -->
{{#each (groupBy items 'FieldOnWhichToGroupBy')}}
    <h1>Group key : {{key}}</h1>
    
    <h1>Group items : </h1>
    {{#each items}}
        <div class="item">
            <ul>
               <li>{{Field1.textValue}}</li>
               <li>{{Field2.textValue}}</li>
           </ul>
        </div>
    {{/each}}
{{/each}}
```

**Comment by MacBoiler** [comment]
This is great Simon-Pierre. I was really sticking on trying to implement my own helper. Got this up and running in less than 10 minutes thanks to you.

Cheers and thanks again.

**Comment by MacBoiler** [comment]
Also, if anyone else is trying to do a similar modification to the CQWP make sure you remove the :f:/r from the SharePoint generated extension URL. This almost got me...:)

**Comment by MacBoiler** [comment]
@spplante I'm having some issues with IE11 (*sigh*) and your CQWP...basically it won't load the helper groupBy JS file. Works on Chrome, Edge no issues.

![image](https://user-images.githubusercontent.com/17294316/67377005-5df34880-f5c8-11e9-97a7-a37384872754.png)

Any thoughts?

PS. This really is an awesome webpart and I just noticed you are the main author. Great work.

**Comment by spplante** [comment]
Hmm looks like the script has not been loaded, do you see any errors in IE’s console?

**Comment by MacBoiler** [comment]
Simon-Pierre, IE11 is throwing CORS errors but the domain is the same for both external file and SharePoint tenant.

Origin https://mytenant.sharepoint.com not found in Access-Control-Allow-Origin header.

I have the deploy resources hosted in an SP library with same domain and no issues. Also, I have a handlebar template accessing same data without a group by and it works fine in IE11.

Particular page has 5 of your CQWP's and the console has 5 of the following:

- Origin https://mytenant.sharepoint.com not found in Access-Control-Allow-Origin header
- XMLHttpRequest: Network Error 0x80700013, Could not complete the operation due to error 80700013.

Also showing [object error] but this is due to not acccessing the JS file.

I'm looking to integrate the groupby in the solution...any thoughts on this?

PS. Really appreciate your help on this.

**Comment by spplante** [comment]
Hmm are you using the 1.4.1 version of the WebPart or the original version (1.*)? I might have an idea of what's going on if you are using the original version.

Otherwise, implementing the group by within the solution is definitely doable, you could do just like the helper and return a built in {{groups}} token on which we could iterate in the handlebars template, the "challenge" is to support grouping by all the different types of fields, but otherwise it should be fairly simple to modify the solution accordingly.

**Comment by MacBoiler** [comment]
@spplante Sorry for being a nube but I'm always amazed at how much I don't know...anyways...are you referring to this...@microsoft/generator-sharepoint@1.7.1...or something else. I have a recent download of your webpart but not sure where to find the webpart version number.

**Comment by spplante** [comment]
When downloading my package, you get the 1.1.0 version (see the version of Microsoft packages in the package.json) which needs you build the WebPart, upload the sources to a "Public CDN enabled" SharePoint library and then store the CDN url in the **write-manifest.json** file. If you use this version but you did NOT upload the sources to a public CDN, it means that your **write-manifest.json** is pointing to my own CDN which is not what you want in production because if I decide to shutdown my SharePoint tenant your WebPart won't work anymore.

See [this article](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/get-started/hosting-webpart-from-office-365-cdn) for the original way of deploying SPFx components.

Starting at version 1.4, SPFx introduced the **includeClientSideAssets** which allows to store the sources directly within the .sppkg, instead of having the .sppkg refer a CDN url. This is now the default setting and in my opinion by far a better option. This is why I included a version of the WebPart which uses 1.4.1 packages at the root of the projet, in the **content-query-webpart-1.4.1.zip**

So long story short, if you did not use the **content-query-webpart-1.4.1.zip** project, you are using the "old" version which uses a .sppkg that points to a CDN url, which would explain the CORS issue because the code does NOT run on your tenant, it runs on mine, and you are trying to refer javascript files from your tenant :)

**Comment by MacBoiler** [comment]
@spplante all of what you're saying makes sense so I'm going to bed now (late Australia time) and in the morning (with a fresh head) I will follow your instructions and then let you know how it works out.

Your help is greatly appreciated!

**Comment by MacBoiler** [comment]
@spplante Just had a quick look an I did change the write-manifest.json to point to my cdnBasePath and this works fine as the webpart will load in IE11 without the groupBy reference in handlebar template.

**Comment by MacBoiler** [comment]
@spplante I tried the includeClientSideAssets and I was getting errors...will look at this again tomorrow and I agree that this would fix it. Only issue would be performance related with multiple webparts on same page.

**Comment by spplante** [comment]
Ok so you do use your own CDN which is good news, but this is most likely what causes the error because your code runs on the CDN url which is different from your tenant URL.

The **includeClientSideAssets** switch only works with the 1.4.1 package, you have to extract the zip to a whole new folder, then build that folder instead. That will generate you a .sppkg with the sources built into it.

**Comment by MacBoiler** [comment]
@spplante I've found the 1.4.1 package...

![image](https://user-images.githubusercontent.com/17294316/67398028-6f028080-f5ed-11e9-97eb-104d95308efd.png)

I think we're getting closer to the solution. Will let you know how I go. Cheers!

**Comment by MacBoiler** [comment]
@spplante I've been working on this all morning and I think there was a couple of issues that were converging and one was the CORS which appears to be related to a MS data centre being out of action...anyways I enabled accessing data sources across domains in ie11 and that has removed the CORS errors and now I can see a syntax error...

![image](https://user-images.githubusercontent.com/17294316/67445293-b7528a80-f64f-11e9-81ba-b07b9e434db1.png)

Looks like ie11 doesn't like the external JS content and this is what I have (with comments removed)...

> ReactContentQuery.ExternalScripts.CQWPGroupByHandlebarExtension = {
>     onPreRender: function(wpContext, handlebarsContext) {
>         handlebarsContext.registerHelper('groupBy', function(items, field, options) {
>             var groupedItems = [];
>             
>             items.forEach(function(item) {
>                 var groupKey = ReactContentQuery.ExternalScripts.CQWPGroupByHandlebarExtension.getGroupKey(item, field);
>                 var existingGroups = groupedItems.filter((group) => { return group.key == groupKey });
>         
>                 if(existingGroups.length > 0) {
>                     existingGroups[0].items.push(item);
>                 }
>                 else {
>                     groupedItems.push({ key: groupKey, items: [item] });
>                 }
>             }); 
> 
>             return groupedItems;
>         });
>     },
>     getGroupKey: function(item, field) {
>         if(!item[field]) {
>             return null;
>         }
>         else if(!isNaN(new Date(item[field].rawValue))) {
>             var objDate = new Date(item[field].rawValue);
>             objDate.setHours(0,0,0,0);
>             return objDate.toISOString();
>         }
>         else if(item[field].rawValue.Url) {
>             return item[field].rawValue.Url;
>         }
>         else {
>             return item[field].textValue;
>         }
>     }
> }

Let me know what you think? I will keep working on this today. My job depends on it. Damn you ie11.

Cheers

MacBoiler

**Comment by MacBoiler** [comment]
@spplante I've worked it out...:)

ie11 does not like es6 arrow functions etc. and only an issue because the external script had not been transpiled.

Converted your external script to es5 and declared ReactContentQuery at top of code...see below...

> var ReactContentQuery;
>
> ReactContentQuery.ExternalScripts.CQWPGroupByHandlebarExtension = {
>   onPreRender: function onPreRender(wpContext, handlebarsContext) {
>     handlebarsContext.registerHelper('groupBy', function (items, field, options) {
>       var groupedItems = [];
>       items.forEach(function (item) {
>         var groupKey = ReactContentQuery.ExternalScripts.CQWPGroupByHandlebarExtension.getGroupKey(item, field);
>         var existingGroups = groupedItems.filter(function (group) {
>           return group.key == groupKey;
>         });
> 
>         if (existingGroups.length > 0) {
>           existingGroups[0].items.push(item);
>         } else {
>           groupedItems.push({
>             key: groupKey,
>             items: [item]
>           });
>         }
>       });
>       return groupedItems;
>     });
>   },
>   getGroupKey: function getGroupKey(item, field) {
>     if (!item[field]) {
>       return null;
>     } else if (!isNaN(new Date(item[field].rawValue))) {
>       var objDate = new Date(item[field].rawValue);
>       objDate.setHours(0, 0, 0, 0);
>       return objDate.toISOString();
>     } else if (item[field].rawValue.Url) {
>       return item[field].rawValue.Url;
>     } else {
>       return item[field].textValue;
>     }
>   }
> };

Thanks for all your help spp. Greatly appreciated.

**Comment by spplante** [comment]
Ah, should have thought about that, I am too used to code in typescript hehe good catch

**Comment by gristy58** [comment]
this is amazing thank you very much.

With a people column i.e. Author, the date function is returned, i assume that is a bug in the example.

**Comment by spplante** [comment]
The code I provided only supports grouping on text, datetime and url fields, you can provide your own implementation to support people field it should be quite straight forward, just add another **if** to the **getGroupKey** function.

**Comment by gristy58** [comment]
thanks i will figure it out, just wanted to say again this is so useful thanks for the great work, have anything else in the pipeline?

**Comment by spplante** [comment]
Not yet, waiting for the SPFx AppSource integration hehe :)

**Comment by salfej** [comment]
This group by function works great! Would you know of a way to sort elements nested within each group? I have to use the available "order by" for the main groups, and I think the sub-groups default sort on creation date, but if I could sort things alphabetically, it would be perfect! Any idea for a solution?

I'm very new to handlebars. 😅

**Comment by spplante** [comment]
@salfej I would modify the implementation of the **groupBy** helper above and specificy another **sortField** parameter which you could use to sort the **groupedItems** variables before returning it using javascript.

You could do something like this for sorting group items by **Title** using javascript's **sort** function with a comparer function :

```javascript
groupedItems.forEach(group => {
    group.items = group.items.sort((a,b) => { 
        return (a.Title > b.Title ? 1 : -1); 
    });
});
```

Didn't test it but should be along those lines!

**Comment by salfej** [comment]
@spplante It's perfect! All I had to change was `a[sortField].textValue` and it worked perfectly! Thanks! 😄

**Comment by kotha1v** [comment]
I'm having some issues, basically it won't load the helper groupBy JS file. 

An Error occurred while processing the handle bar template, 
Error 404 for the JS file


Any thoughts?

**Comment by spplante** [comment]
You are sure the url of your file is correct? 404 usually means your file is not found. If the JS file is stored in SharePoint, make sure it is checked in and published.

**Comment by hugoabernier** [comment]
@MacBoiler did you ever solve this issue? It looks like @spplante had some great suggestions to solve this.

I'll mark the issue as closed, but feel free to re-open it if you need more help. I also  encourage you to take the time to fill the "big form" in the future, as it makes it easier for other community members to help you.

Thanks @spplante for your help.

**Comment by MacBoiler** [comment]
@hugoabernier yes this issue was resolved and big shoutout to @spplante for all his help.

Also, apologies for my "big form" comment...I was under a bit of pressure and had no sleep. I will fill out the big form in the future.:)
